[2.7] bpo-31170: Update libexpat from 2.2.3 to 2.2.4 (#3315)

vstinner opened the pull request:
* bpo-31170: Update libexpat from 2.2.3 to 2.2.4

Fix copying of partial characters for UTF-8 input (libexpat bug 115):
https://github.com/libexpat/libexpat/issues/115

* Add NEWS entry.

(cherry picked from commit 759e30ec47048cb9835c62aaeac48748c8151390)

<!-- issue-number: bpo-31170 -->
https://bugs.python.org/issue31170
<!-- /issue-number -->

vstinner committed:
bpo-31170: Update libexpat from 2.2.3 to 2.2.4 (#3315)

* bpo-31170: Update libexpat from 2.2.3 to 2.2.4

Fix copying of partial characters for UTF-8 input (libexpat bug 115):
https://github.com/libexpat/libexpat/issues/115

* Add NEWS entry.

(cherry picked from commit 759e30ec47048cb9835c62aaeac48748c8151390)

vstinner committed:
Merge branch '2.7' into expat224_27

tiran reviewed (dismissed):
Breaks Windows builds:

```
..\Modules\expat\xmltok.c(34): fatal error C1083: Cannot open include file: 'stdbool.h': No such file or directory [C:\projects\cpython\PCbuild\pyexpat.vcxproj]
```

vstinner committed:
expat: Fix compile error with Visual Studio 2012

The standard header stdbool.h is not available
with old Visual Studio compilers

Cherry-picked from libexpat b4b89c2ab0cc5325a41360c25ef9d2ccbe617e5c.

vstinner commented:
"Breaks Windows builds: ..\Modules\expat\xmltok.c(34): fatal error C1083: Cannot open include file: 'stdbool.h': No such file or directory [C:\projects\cpython\PCbuild\pyexpat.vcxproj]"

I spent 15 minutes to write a fix for libexpat. But i was working on the 2.2.4 branch, in fact the bug is already fixed upstream: https://github.com/libexpat/libexpat/commit/b4b89c2ab0cc5325a41360c25ef9d2ccbe617e5c

So I cherry-picked the fix.

vstinner committed:
expat: Add artificial scopes in xmltok.c utf8_toUtf8() to fix c89 compilation.

Cherry-picked from libexpat commit e0b290eb3d8f4c4b45137a7d7f4f8db812145bd2

benjaminp commented:
Thanks. Since this is going into 2.7.14 final, I'll deal with it.

benjaminp commented:
c00d5087cd308cc8be41e0afd8df27726185347f
